src/storage/src/hummock/utils.rs

```diff
@@ -166,23 +170,88 @@ pub fn prune_nonoverlapping_ssts<'a>(
     ssts[start_table_idx..=end_table_idx].iter()
 }
 
-#[derive(Debug)]
+type RequestQueue = VecDeque<(Sender<MemoryTrackerImpl>, u64)>;
+enum MemoryRequest {
+    Ready(MemoryTrackerImpl),
+    Pending(Receiver<MemoryTrackerImpl>),
+}
+
+pub struct PendingRequestCancelGuard {
+    inner: Option<Arc<MemoryLimiterInner>>,
+    rx: Receiver<MemoryTrackerImpl>,
+}
+
+impl Drop for PendingRequestCancelGuard {
+    fn drop(&mut self) {
+        if let Some(limiter) = self.inner.take() {
+            // close rc before calling `try_recv`, it will make `MemoryTrackerImpl` which generates after dropping of `PendingRequestCancelGuard` drop in loop of `may_notify_waiters` in other thread.
+            // If `MemoryTrackerImpl` send before this thread calling `close`, it can still be received by this thread. Once this thread receives the message, it need drop the message and update `total_size` in `MemoryTrackerImpl`'s drop.
+            self.rx.close();
+            if let Ok(msg) = self.rx.try_recv() {
+                drop(msg);
+                // every time `MemoryTrackerImpl` drop, it will update `total_size` and we need to check whether there exist waiters to be notified.
+                limiter.may_notify_waiters();
+            }
+        }
+    }
+}
+
+impl Future for PendingRequestCancelGuard {
+    type Output = Option<MemoryTrackerImpl>;
+
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Self::Output> {
+        match self.rx.poll_unpin(cx) {
+            Poll::Ready(Ok(msg)) => {
+                self.inner.take();
+                Poll::Ready(Some(msg))
+            }
+            Poll::Ready(Err(_)) => {
+                self.inner.take();
+                Poll::Ready(None)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
 struct MemoryLimiterInner {
     total_size: AtomicU64,
-    notify: Notify,
+    controller: Mutex<RequestQueue>,
+    has_waiter: AtomicBool,
     quota: u64,
 }
 
 impl MemoryLimiterInner {
     fn release_quota(&self, quota: u64) {
-        self.total_size.fetch_sub(quota, AtomicOrdering::Release);
-        self.notify.notify_waiters();
+        self.total_size.fetch_sub(quota, AtomicOrdering::SeqCst);
     }
 
     fn add_memory(&self, quota: u64) {
         self.total_size.fetch_add(quota, AtomicOrdering::SeqCst);
     }
 
+    fn may_notify_waiters(self: &Arc<Self>) {
+        // check `has_waiter` to avoid access lock every times drop `MemoryTracker`.
+        if !self.has_waiter.load(AtomicOrdering::Acquire) {
+            return;
+        }
+        let mut waiters = self.controller.lock();
+        while let Some((_, quota)) = waiters.front() {
+            if !self.try_require_memory(*quota) {
+                break;
+            }
+            let (tx, quota) = waiters.pop_front().unwrap();
+            let _ = tx.send(MemoryTrackerImpl::new(self.clone(), quota));
```

Comment:
The `MemoryTrackerImpl` and the newly added `PendingRequestCancelGuard` seem to be an over-design for the corner case mentioned above. Let's analyze all the cases when quota is acquired:
1. When the quota is acquired via `MemoryRequest::Ready` (L287), we need to release the quota and notify waiters.
2. When the quota is acquired here in `may_notify_waiters` in L243, there are two sub-cases:
    a. `tx.send` here succeeds, regardless of whether the rx is dropped or not, we need to release quota and notify waiter.
    b. `tx.send` here fails (due to rx has been dropped before tx.send). we only need to release quota but **not** notify watier (Otherwise, it will deadlock).


Therefore, the implementation can be greatly simplified without introducing `MemoryTrackerImpl` and `PendingRequestCancelGuard`.

```rust
// MemoryTrackerImpl is not needed.
struct MemoryTracker {
    limiter: Arc<MemoryLimiterInner>,
    quota: Option<u64>,
}

// MemoryTracker::drop will release quota + notifier waiters if quota is not None (1 + 2.a)
impl Drop for MemoryTracker {
    fn drop(&mut self) {
        if let Some(quota) = self.quota.take() {
            self.limiter.release_quota(quota);
            self.limiter.may_notify_waiters();
        }
    }
}

// Add a new method to releae quota without notify waiters.
impl MemoryTracker {
  fn release_quota(&mut self) {
      if let Some(quota) = self.quota.take() {
            self.limiter.release_quota(quota);
      }
  }
}

...

// MemoryRequest takes MemoryTracker
enum MemoryRequest {
    Ready(MemoryTracker),
    Pending(Receiver<MemoryTracker>),
}

...

impl MemoryLimiterInner {
    ...
    fn may_notify_waiters(self: &Arc<Self>) {
        ...
        let mut waiters = self.controller.lock();
        while let Some((_, quota)) = waiters.front() {
            if !self.try_require_memory(*quota) {
                break;
            }
            ...
            if let Err(SendError(tracker)) = tx.send(MemoryTracker::new(self.clone(), quota)) {
                // 2.b
                tracker.release_quota(quota);
            }
    }
    ...
}

...

impl MemoryLimiter {
    pub async fn require_memory(&self, quota: u64) -> MemoryTracker {
        // PendingRequestCancelGuard is not needed
        match self.inner.require_memory(quota) {
            MemoryRequest::Ready(tracker) => tracker,
            MemoryRequest::Pending(rx) => rx.await
    }
}
```

Reply:
We can not send `MemoryTracker` with in lock of  `controller` because if it drops in this method, it may call `may_notify_waiters ` again, which would acquire one lock twice.

Reply:
We met this bug in https://github.com/risingwavelabs/risingwave/pull/6634

Reply:
If https://github.com/tokio-rs/tokio/pull/6558 looks good from the tokio community, we can have a patch on the tokio version we are using, and then we can adopt the code in this comment.